### PR TITLE
Decouple categories/states from folder location (#386)

### LINKED ISSUE #386: Decouple categories/states from folder location - user-defined categories with optional folder mapping
## Summary

Currently task state/category is tightly coupled to folder location (e.g. `priority/`, `backlog/`, `done/`). This should be decoupled so users can manually define their own categories and states, with folder mapping being optional rather than required.

## Desired behaviour

- Users define categories/states in settings (name, display order, colour, etc.)
- Each category can optionally map to a folder path, but doesn't have to
- State can be determined by other mechanisms: frontmatter field, filename pattern, or purely manual assignment
- Moving a task between columns updates the state via the configured mechanism (frontmatter write, folder move, or both)
- Existing folder-based behaviour remains available as one strategy, not the only one

## Motivation

- Different vaults have different organisational structures - not everyone wants state encoded as folder location
- Enables use cases like tags-as-state, frontmatter-as-state, or hybrid approaches
- Reduces friction when adapting the plugin to existing vault layouts
- Unlocks virtual categories (see #385) more naturally

## Implementation notes

- This is a significant adapter-level change - the task-agent adapter currently hardcodes `STATE_FOLDER_MAP`
- The framework's `ItemMover` interface may need to become more abstract (move = state transition, not necessarily file move)
- Consider a migration path for existing users whose state is folder-derived
- Could be modelled as a "state resolver" strategy pattern: folder-based, frontmatter-based, manual, or composite

### PR DESCRIPTION
**Pull request by tomcorke**
## Summary

Introduces a StateResolver strategy pattern that decouples task state from folder location. Users can now choose how state is determined:

- **Folder-based** (default) - existing behavior, derives state from folder path
- **Frontmatter-based** - reads/writes a `state` frontmatter field
- **Composite** - reads frontmatter first, falls back to folder

### Changes

- `StateResolver` interface in `src/core/interfaces.ts` with `resolveState()` and `applyState()` methods
- Three concrete resolvers: `FolderStateResolver`, `FrontmatterStateResolver`, `CompositeStateResolver`
- `TaskParser` refactored to use resolver instead of hardcoded folder lookup
- `TaskMover` refactored to delegate state transitions to resolver's `applyState()`
- Settings UI for choosing resolution strategy in adapter config
- Full backward compatibility - folder-based is default, zero-config upgrade

### Test plan

- [ ] Existing folder-based workflows unchanged (regression)
- [ ] Frontmatter strategy reads/writes `state` field correctly
- [ ] Composite strategy falls back from frontmatter to folder
- [ ] Settings UI reflects strategy choice
- [ ] Drag-drop between columns uses selected strategy

Closes #386

### CODE REVIEW

File: src/core/resolvers/FrontmatterStateResolver.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  ): Promise<boolean> {
+    const content = await app.vault.read(file);
+    const fieldPattern = new RegExp(`^${escapeRegex(this.fieldName)}:\\s*.+$`, "m");
+    if (!fieldPattern.test(content)) return false;
+
+    const updated = content.replace(fieldPattern, `${this.fieldName}: ${newState}`);
+    await app.vault.modify(file, updated);
+    return true;
+  }
+
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/core/resolvers/FrontmatterStateResolver.ts]: Fixed. `applyState` now upserts the field within the `--- ... ---` frontmatter block: if the field exists (including blank values like `state:`) it is updated in place; if absent it is inserted at the end of the frontmatter block. The regex is now scoped to the frontmatter block only, so it cannot match content in the document body. Returns false only when there is no frontmatter block at all.

File: src/core/resolvers/FolderStateResolver.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    newState: string,
+    _oldState: string,
+    basePath: string,
+  ): Promise<boolean> {
+    const targetFolder = this.stateToFolder[newState];
+    if (!targetFolder) return false;
+
+    const newFolderPath = `${basePath}/${targetFolder}`;
+    const newPath = `${newFolderPath}/${file.name}`;
+
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/core/resolvers/FolderStateResolver.ts:66]: Fixed. `applyState` now normalizes `basePath` by stripping trailing slashes, and handles empty `basePath` by using `targetFolder` directly (no leading slash). Added tests for both edge cases.

File: src/adapters/task-agent/TaskParser.ts [hunk 3]
```diff
@@ -260,6 +263,25 @@ export class TaskParser implements WorkItemParser {
...
+    path: string,
+    frontmatter: Record<string, unknown> | undefined,
+  ): TaskState | null {
+    if (this.stateResolver) {
+      const resolved = this.stateResolver.resolveState(path, frontmatter);
+      if (resolved !== null && VALID_STATES.includes(resolved as TaskState)) {
+        return resolved as TaskState;
+      }
+      // If resolver returned a value but it's not a valid TaskState, fall through
+      if (resolved !== null) return null;
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/adapters/task-agent/TaskParser.ts]: Fixed. When the resolver returns a non-null value that is not in `VALID_STATES`, the parser now falls through to `getStateFromPath()` instead of returning `null`. Added a test verifying that tasks with unrecognized resolver states are resolved via folder path fallback.

File: src/adapters/task-agent/stateResolverFactory.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * work, while tasks with a `state` frontmatter field use that value.
+ *
+ * The folder resolver handles file moves on state transition. The frontmatter
+ * resolver handles the `state:` field update (which TaskMover also does
+ * directly for now - the resolver is used for state *resolution*, the mover
+ * handles the full transition including tags, timestamps, and activity log).
+ */
+export function createDefaultStateResolver(basePath: string): StateResolver {
+  return createStateResolver("folder", basePath);
+}
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/adapters/task-agent/stateResolverFactory.ts:21]: Fixed. Updated the docstring to accurately describe the implementation: folder-based resolution is the default for backward compatibility. The old docstring incorrectly described composite behavior.

File: src/adapters/task-agent/TaskMover.ts [hunk 5]
```diff
@@ -64,17 +70,24 @@ export class TaskMover implements WorkItemMover {
...
-      const newPath = `${newFolderPath}/${file.name}`;
-
-      if (file.path !== newPath) {
-        // Ensure target folder exists
-        const folder = this.app.vault.getAbstractFileByPath(newFolderPath);
-        if (!folder) {
-          await this.app.vault.createFolder(newFolderPath);
+      // Apply the state transition (file move or other mechanism)
+      if (this.stateResolver) {
+        await this.stateResolver.applyState(this.app, file, newColumn, oldState, this.basePath);
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/adapters/task-agent/TaskMover.ts]: Fixed. `move()` now checks the boolean return value of `stateResolver.applyState()` and returns `false` when the resolver fails, so the UI can surface the error instead of reporting success. Added tests for both success and failure paths.

File: src/adapters/task-agent/TaskAgentConfig.ts [hunk 6]
```diff
@@ -34,6 +34,19 @@ export const TASK_AGENT_CONFIG: PluginConfig = {
...
+      name: "State resolution strategy",
+      description:
+        "How task state is determined. Folder: derived from folder location (default). Frontmatter: read from the state field. Composite: check frontmatter first, fall back to folder.",
+      type: "dropdown",
+      default: "folder",
+      choices: {
+        folder: "Folder-based (default)",
+        frontmatter: "Frontmatter field",
+        composite: "Composite (frontmatter + folder fallback)",
+      },
```
[the remark this answers is not in the document]
Comment (tomcorke) [reply src/adapters/task-agent/TaskAgentConfig.ts:48]: Fixed. Updated the setting description to clarify that 'Frontmatter' falls back to folder location when the field is missing, and that 'Composite' applies both mechanisms on state transition. This matches the actual behavior in `TaskParser.resolveStateFromFile()`.
